File: java_template/src/main/java/lambda/CSVProcessor.java
package lambda;


import java.io.ByteArrayInputStream;
import java.io.InputStream;


import java.io.StringWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import org.apache.commons.csv.CSVFormat;

import org.apache.commons.csv.CSVPrinter;


import saaf.Inspector;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;

// //imorts for aws s3

import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.S3Object;

import java.util.HashMap;
import java.util.Scanner;

/*
Example Service #1 transformations (can implement others):
 1. Add column [Order Processing Time] column that stores an integer value representing the number of days between the [Order Date] and [Ship Date]

 2. Transform [Order Priority] column: L to “Low” M to “Medium” H to “High” C to “Critical”

 3. Add a [Gross Margin] column. The Gross Margin Column is a percentage calculated using the formula: [Total Profit] /  [Total Revenue].
 It is stored as a floating point value (e.g 0.25 for 25% profit).

 4. Remove duplicate data identified by [Order ID]. Any record having a duplicate [Order ID] that has already been processed will be ignored.
 */

public class CSVProcessor implements RequestHandler<Request, HashMap<String, Object>> {

    public HashMap<String, Object> handleRequest(Request request, Context context) {
        String filename = "transformed.csv";
        Inspector inspector = new Inspector();
        // inspector.inspectAll();

        String srcBucket = "test.bucket.462562f23.bm";
        String srcKey = "data.csv";

        AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();
        S3Object s3Object = s3Client.getObject(new GetObjectRequest(
                srcBucket, srcKey));


        InputStream objectData = s3Object.getObjectContent();

        try {

            Scanner scanner = new Scanner(objectData);

            // // Read CSV data from S3
            List<List<String>> recordsList = new ArrayList<>();

            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                String[] values = line.split(",");
                List<String> record = new ArrayList<>();
                for (String value : values) {
                    record.add(value);
                }
                recordsList.add(record);
            }

            transformOrderPriority(recordsList);
            addOrderProcessingTime(recordsList);
            addGrossMargin(recordsList);
            removeDuplicateData(recordsList);

            // Write the transformed data to a StringWriter
            StringWriter writer = new StringWriter();
            try (CSVPrinter printer = new CSVPrinter(writer, CSVFormat.DEFAULT)) {
                for (List<String> record : recordsList) {
                    printer.printRecord(record);
                }
            }

            scanner.close();
            LambdaLogger logger = context.getLogger();
            logger.log("ProcessCSV bucketname:" + srcBucket + " filename:" + srcKey + "\n");

            try {

                byte[] bytes = writer.toString().getBytes(StandardCharsets.UTF_8);
                InputStream is = new ByteArrayInputStream(bytes);
                ObjectMetadata meta = new ObjectMetadata();
                meta.setContentLength(bytes.length);
                meta.setContentType("text/plain"); // Create new file on S3

                s3Client.putObject(srcBucket, filename, is, meta);
                System.out.println("Succesfully uploaded transformed file to S3");

                          Response response = new Response();
            response.setValue("Bucket:" + srcBucket + " filename:" + filename);

            // inspector.consumeResponse(response);
                return inspector.finish();

            } catch (Exception e) {
                e.printStackTrace();
                System.out.println("Error uploading file to S3");
                return inspector.finish();
            }
        }
        catch (Exception e) {
            e.printStackTrace();
            System.out.println("Error reading CSV from S3");
            return inspector.finish();
        }

    }

    private static void addOrderProcessingTime(List<List<String>> recordsList) {
        recordsList.get(0).add("Processing Time");

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("M/d/yyyy");

        // Skip the header row
        for (int i = 1; i < recordsList.size(); i++) {
            List<String> record = recordsList.get(i);

            LocalDate orderDate = LocalDate.parse(record.get(5), formatter);
            LocalDate shipDate = LocalDate.parse(record.get(7), formatter);
            long orderProcessingTime = java.time.temporal.ChronoUnit.DAYS.between(orderDate, shipDate);

            record.add(String.valueOf(orderProcessingTime));
        }
    }

    private static void transformOrderPriority(List<List<String>> recordsList) {
        // Skip the header row
        for (int i = 1; i < recordsList.size(); i++) {
            List<String> record = recordsList.get(i);
            String orderPriority = record.get(4);

            switch (orderPriority) {
                case "L":
                    orderPriority = "Low";
                    break;
                case "M":
                    orderPriority = "Medium";
                    break;
                case "H":
                    orderPriority = "High";
                    break;
                case "C":
                    orderPriority = "Critical";
                    break;
            }
            record.set(4, orderPriority); // Modify the existing Order Priority column
        }
    }

    // Add a [Gross Margin] column. The Gross Margin Column is a percentage
    // calculated using the formula: [Total Profit] / [Total Revenue].
    // It is stored as a floating point value (e.g 0.25 for 25% profit).
    private static void addGrossMargin(List<List<String>> recordsList) {
        recordsList.get(0).add("Gross Margin");
        // Skip the header row
        for (int i = 1; i < recordsList.size(); i++) {
            List<String> record = recordsList.get(i);

            float totalProfit = Float.parseFloat(record.get(13));
            float totalRevenue = Float.parseFloat(record.get(11));
            float grossMargin = totalProfit / totalRevenue;

            record.add(String.valueOf(grossMargin)); // Add the new column
        }
    }

    // Remove duplicate data identified by [Order ID]. Any record having a duplicate
    // [Order ID] that has already been processed will be ignored.
    private static void removeDuplicateData(List<List<String>> recordsList) {
        ArrayList<String> processedOrderIds = new ArrayList<>();

        // Skip the header row
        for (int i = 1; i < recordsList.size(); i++) {
            List<String> record = recordsList.get(i);
            String orderId = record.get(6);
            if (processedOrderIds.contains(orderId)) {
                recordsList.remove(i);
            } else {
                processedOrderIds.add(orderId);
            }
        }
    }
}

File: python_template/src/CSVProcessor.py
import csv
import io
import datetime
import boto3
import uuid
import Inspector
from botocore.exceptions import NoCredentialsError

class CSVProcessor:
    @staticmethod
    def add_order_processing_time(records_list):
        records_list[0].append("Processing Time")
        date_format = "%m/%d/%Y"
        
        for record in records_list[1:]:
            order_date = datetime.datetime.strptime(record[5], date_format)
            ship_date = datetime.datetime.strptime(record[7], date_format)
            order_processing_time = (ship_date - order_date).days
            record.append(str(order_processing_time))

    @staticmethod
    def transform_order_priority(records_list):
        for record in records_list[1:]:
            order_priority = record[4]

            if order_priority == "L":
                record[4] = "Low"
            elif order_priority == "M":
                record[4] = "Medium"
            elif order_priority == "H":
                record[4] = "High"
            elif order_priority == "C":
                record[4] = "Critical"


    # Add a [Gross Margin] column. The Gross Margin Column is a percentage
    # calculated using the formula: [Total Profit] / [Total Revenue].
    # It is stored as a floating point value (e.g 0.25 for 25% profit).
    @staticmethod
    def add_gross_margin(records_list):
        records_list[0].append("Gross Margin")

        for record in records_list[1:]:
            total_profit = float(record[13])
            total_revenue = float(record[11])
            gross_margin = total_profit / total_revenue
            record.append(str(gross_margin))


    # Remove duplicate data identified by [Order ID]. Any record having a duplicate
    # [Order ID] that has already been processed will be ignored.
    @staticmethod
    def remove_duplicate_data(records_list):
        processed_order_ids = set()
        index_to_remove = []

        # Skip the header row
        for i in range(1, len(records_list)):
            record = records_list[i]
            order_id = record[6]

            if order_id in processed_order_ids:
                index_to_remove.append(i)
            else:
                processed_order_ids.add(order_id)

        # Remove duplicate rows
        for index in reversed(index_to_remove):
            records_list.pop(index)


def handle_request(event, lambda_context):
    try:
        # Extract values from the event (event is the input to the Lambda function)
        src_bucket = event['src_bucket']
        src_key = event['src_key']
        filename = "transformed.csv"
        s3 = boto3.client('s3')

        response = s3.get_object(Bucket=src_bucket, Key=src_key)
        object_data = response['Body'].read().decode('utf-8')
        records_list = list(csv.reader(io.StringIO(object_data)))

        CSVProcessor.transform_order_priority(records_list)
        CSVProcessor.add_order_processing_time(records_list)
        CSVProcessor.add_gross_margin(records_list)
        CSVProcessor.remove_duplicate_data(records_list)


        # Write the transformed data to a new CSV file
        transformed_data = io.StringIO()
        csv.writer(transformed_data).writerows(records_list)

        # Upload the transformed data to S3
        s3.put_object(Body=transformed_data.getvalue(), Bucket=src_bucket, Key=filename)
        print("Successfully uploaded transformed file to S3")

        return {"Bucket": src_bucket, "Filename": filename}

    except NoCredentialsError:
        print("Credentials not available")
        return {}